{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63fef4dd-4757-4484-b807-adea45cadb77",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import random"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6b7c0722-a734-4077-a9cf-565fbfc649b6",
   "metadata": {},
   "source": [
    "#turn this cell to code if needed\n",
    "#you will need bcftools to run the script. Make sure to be able to run it \n",
    "#you might need to import the module with a command like this\n",
    "\n",
    "\n",
    "exec(open(\"/shared/software/modules/4.6.1/init/python.py\").read(), globals())\n",
    "module(\"load\", \"bcftools\")\n",
    "print(\"bcftools has been loaded\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc95d1c-851a-4b30-aceb-21809068cd92",
   "metadata": {},
   "source": [
    "# Creating the reference.smap from the igsr_samples.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "27e796bd-fe50-4adf-9243-bb771a1b05ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#since the same sample are in the three different file we can choose any file we want here\n",
    "reference_file='../data/input/real/65k_SNP/65kSNP_chr1.vcf.gz'\n",
    "\n",
    "legend_file='../data/input/real/igsr_samples.tsv'\n",
    "reference_map_file='../data/input/real/reference.smap'\n",
    "\n",
    "#creating the .smap file\n",
    "sample_list=pd.read_csv(legend_file,sep=\"\\t\")\n",
    "reference_map= sample_list.loc[:, ['Sample name', 'Superpopulation code']]\n",
    "\n",
    "#deleting Frow if superpopulation code is Nan\n",
    "reference_map = reference_map.dropna(subset=['Superpopulation code'])\n",
    "\n",
    "#removing the EUR,AFR individual\n",
    "reference_map=reference_map[reference_map.iloc[:,1]!='EUR,AFR']\n",
    "\n",
    "sample_name = subprocess.check_output(['bcftools', 'query', '-l', reference_file], universal_newlines=True)\n",
    "sample_name=list(sample_name.strip().split('\\n'))\n",
    "\n",
    "#deleting row if the sample is not in the reference file\n",
    "mask= reference_map['Sample name'].isin(sample_name)\n",
    "reference_map = reference_map[mask]\n",
    "\n",
    "np.savetxt(reference_map_file,reference_map,delimiter='\\t',fmt=\"%s\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c012065-a84a-426c-b7a1-036a4f7c85ad",
   "metadata": {},
   "source": [
    "## Extracting specific subpopulations from igsr file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8772e772-7706-4ca6-80d2-49024c70bb5a",
   "metadata": {},
   "source": [
    "In this example we want to extract PUR subpopulation from the referece vcf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "777675c2-3f86-4a2f-b658-832068723e22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Population code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00290</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00308</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>NA21143</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>NA21112</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>NA21117</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2501</th>\n",
       "      <td>NA21124</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2502</th>\n",
       "      <td>NA21129</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2503 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample name Population code\n",
       "0        HG00271             EUR\n",
       "1        HG00276             EUR\n",
       "2        HG00288             EUR\n",
       "3        HG00290             EUR\n",
       "4        HG00308             EUR\n",
       "...          ...             ...\n",
       "2498     NA21143             SAS\n",
       "2499     NA21112             SAS\n",
       "2500     NA21117             SAS\n",
       "2501     NA21124             SAS\n",
       "2502     NA21129             SAS\n",
       "\n",
       "[2503 rows x 2 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "legend_file='./data/input/real/igsr_samples.tsv'\n",
    "df=pd.read_csv(legend_file,sep='\\t')\n",
    "reference_map_file='./data/input/real/65k_SNP/reference.smap'\n",
    "reference_map=pd.read_csv(reference_map_file,delimiter='\\t',header=None)\n",
    "reference_map.columns=[\"Sample name\",\"Population code\"]\n",
    "reference_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "dc42c742-717f-4321-b010-98e8d7993dad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Biosample ID</th>\n",
       "      <th>Population code</th>\n",
       "      <th>Population name</th>\n",
       "      <th>Superpopulation code</th>\n",
       "      <th>Superpopulation name</th>\n",
       "      <th>Population elastic ID</th>\n",
       "      <th>Data collections</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00271</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME123417</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00276</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME123424</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00288</td>\n",
       "      <td>female</td>\n",
       "      <td>SAME1839246</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00290</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1839057</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38,1000 Genomes 30x on GRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00303</td>\n",
       "      <td>male</td>\n",
       "      <td>SAME1840115</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>FIN</td>\n",
       "      <td>1000 Genomes on GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4973</th>\n",
       "      <td>HGDP00773</td>\n",
       "      <td>female</td>\n",
       "      <td>SAMEA3302906</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Japanese</td>\n",
       "      <td>NaN</td>\n",
       "      <td>East Asia (SGDP)</td>\n",
       "      <td>JapaneseSGDP</td>\n",
       "      <td>Simons Genome Diversity Project</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4974</th>\n",
       "      <td>R3</td>\n",
       "      <td>male</td>\n",
       "      <td>SAMEA3302714</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Relli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>South Asia (SGDP)</td>\n",
       "      <td>RelliSGDP</td>\n",
       "      <td>Simons Genome Diversity Project</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4975</th>\n",
       "      <td>NA12236</td>\n",
       "      <td>female</td>\n",
       "      <td>SAMEA6604124</td>\n",
       "      <td>CEU</td>\n",
       "      <td>CEPH</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European Ancestry</td>\n",
       "      <td>CEU</td>\n",
       "      <td>1000 Genomes 30x on GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4976</th>\n",
       "      <td>HGDP00456</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mbuti</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Africa (HGDP)</td>\n",
       "      <td>MbutiHGDP</td>\n",
       "      <td>HGDP Transcriptome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>GM19129</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African Ancestry</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Human Genome Structural Variation Consortium, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4978 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample name     Sex  Biosample ID Population code Population name  \\\n",
       "0        HG00271    male    SAME123417             FIN         Finnish   \n",
       "1        HG00276  female    SAME123424             FIN         Finnish   \n",
       "2        HG00288  female   SAME1839246             FIN         Finnish   \n",
       "3        HG00290    male   SAME1839057             FIN         Finnish   \n",
       "4        HG00303    male   SAME1840115             FIN         Finnish   \n",
       "...          ...     ...           ...             ...             ...   \n",
       "4973   HGDP00773  female  SAMEA3302906             NaN        Japanese   \n",
       "4974          R3    male  SAMEA3302714             NaN           Relli   \n",
       "4975     NA12236  female  SAMEA6604124             CEU            CEPH   \n",
       "4976   HGDP00456    male           NaN             NaN           Mbuti   \n",
       "4977     GM19129  female           NaN             YRI          Yoruba   \n",
       "\n",
       "     Superpopulation code Superpopulation name Population elastic ID  \\\n",
       "0                     EUR    European Ancestry                   FIN   \n",
       "1                     EUR    European Ancestry                   FIN   \n",
       "2                     EUR    European Ancestry                   FIN   \n",
       "3                     EUR    European Ancestry                   FIN   \n",
       "4                     EUR    European Ancestry                   FIN   \n",
       "...                   ...                  ...                   ...   \n",
       "4973                  NaN     East Asia (SGDP)          JapaneseSGDP   \n",
       "4974                  NaN    South Asia (SGDP)             RelliSGDP   \n",
       "4975                  EUR    European Ancestry                   CEU   \n",
       "4976                  NaN        Africa (HGDP)             MbutiHGDP   \n",
       "4977                  AFR     African Ancestry                   YRI   \n",
       "\n",
       "                                       Data collections  \n",
       "0     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "1     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "2     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "3     1000 Genomes on GRCh38,1000 Genomes 30x on GRC...  \n",
       "4                                1000 Genomes on GRCh38  \n",
       "...                                                 ...  \n",
       "4973                    Simons Genome Diversity Project  \n",
       "4974                    Simons Genome Diversity Project  \n",
       "4975                         1000 Genomes 30x on GRCh38  \n",
       "4976                                 HGDP Transcriptome  \n",
       "4977  Human Genome Structural Variation Consortium, ...  \n",
       "\n",
       "[4978 rows x 9 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7da3a7eb-7a3f-4cac-887e-bcdbe123fa13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Sample name Population code\n",
      "0        HG00271             EUR\n",
      "1        HG00276             EUR\n",
      "2        HG00288             EUR\n",
      "3        HG00290             EUR\n",
      "4        HG00308             EUR\n",
      "...          ...             ...\n",
      "2498     NA21143             SAS\n",
      "2499     NA21112             SAS\n",
      "2500     NA21117             SAS\n",
      "2501     NA21124             SAS\n",
      "2502     NA21129             SAS\n",
      "\n",
      "[2503 rows x 2 columns]\n",
      "\n",
      "\n",
      "101     HG00555\n",
      "109     HG00733\n",
      "111     HG00738\n",
      "113     HG00740\n",
      "119     HG01161\n",
      "         ...   \n",
      "2769    HG01303\n",
      "2772    HG01308\n",
      "2811    HG01396\n",
      "2814    HG01404\n",
      "2817    HG01411\n",
      "Name: Sample name, Length: 150, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample name</th>\n",
       "      <th>Population code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00740</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG01161</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG01173</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG01197</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG01200</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>HG01098</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>HG01101</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>HG01303</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>HG01308</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>HG01396</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample name Population code\n",
       "0       HG00740             AMR\n",
       "1       HG01161             AMR\n",
       "2       HG01173             AMR\n",
       "3       HG01197             AMR\n",
       "4       HG01200             AMR\n",
       "..          ...             ...\n",
       "99      HG01098             AMR\n",
       "100     HG01101             AMR\n",
       "101     HG01303             AMR\n",
       "102     HG01308             AMR\n",
       "103     HG01396             AMR\n",
       "\n",
       "[104 rows x 2 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PUR=df[df[\"Population code\"]=='PUR']\n",
    "sample_PUR=PUR[\"Sample name\"]\n",
    "print(reference_map)\n",
    "print('\\n')\n",
    "print(sample_PUR)\n",
    "\n",
    "sample_PUR=sample_PUR.to_frame().merge(reference_map)\n",
    "sample_PUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "3fb8552f-ce45-41de-9d4d-3d941eb818e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt(\"./data/input/real/PUR_sample.tsv\", sample_PUR['Sample name'], delimiter=\"\\t\", fmt=\"%s\")\n",
    "np.savetxt(\"./data/input/real/PUR_sample.smap\",sample_PUR,delimiter='\\t',fmt='%s')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Mon Python",
   "language": "python",
   "name": "ter"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
